File: Plugins/Android/rudder-sdk-unity-plugin-android/src/main/java/com/rudderlabs/android/sdk/core/RudderConfigBuilder.java
package com.rudderlabs.android.sdk.core;

import android.text.TextUtils;
import android.view.contentcapture.ContentCaptureSession;
import android.webkit.URLUtil;

import java.util.ArrayList;
import java.util.List;

/*
 * builder for RudderConfig
 *
 * isDebug -> true => set the logLevel to DEBUG automatically else sets as defined by the user
 * */
public class RudderConfigBuilder {
    private String endPointUri = Constants.BASE_URL;

    public RudderConfigBuilder withEndPointUri(String endPointUri) throws RudderException {
        if (TextUtils.isEmpty(endPointUri)) {
            throw new RudderException("endPointUri can not be null or empty");
        }
        if (!URLUtil.isValidUrl(endPointUri)) {
            throw new RudderException("malformed endPointUri");
        }
        this.endPointUri = endPointUri;
        return this;
    }

    private int flushQueueSize = Constants.FLUSH_QUEUE_SIZE;

    public RudderConfigBuilder withFlushQueueSize(int flushQueueSize) throws RudderException {
        if (flushQueueSize < 1 || flushQueueSize > 100) {
            throw new RudderException("flushQueueSize is out of range. Min: 1, Max: 100");
        }
        this.flushQueueSize = flushQueueSize;
        return this;
    }

    private boolean isDebug = false;

    public RudderConfigBuilder withDebug(boolean isDebug) {
        this.isDebug = isDebug;
        return this;
    }

    private int logLevel = RudderLogger.RudderLogLevel.NONE;

    public RudderConfigBuilder withLogLevel(int logLevel) {
        this.logLevel = logLevel;
        return this;
    }

    private int dbThresholdCount = Constants.DB_COUNT_THRESHOLD;

    public RudderConfigBuilder withDbThresholdCount(int dbThresholdCount) {
        this.dbThresholdCount = dbThresholdCount;
        return this;
    }

    private int sleepTimeout = Constants.SLEEP_TIMEOUT;

    public RudderConfigBuilder withSleepCount(int sleepCount) {
        this.sleepTimeout = sleepCount;
        return this;
    }

    public RudderConfig build() throws RudderException {
        return new RudderConfig(endPointUri, flushQueueSize, dbThresholdCount, sleepTimeout, isDebug ? RudderLogger.RudderLogLevel.DEBUG : logLevel);
    }
}

File: RudderUnityPluginDemo/Assets/RudderUnityPlugin/RudderConfigBuilder.cs
using System.Collections.Generic;

namespace Rudderlabs
{


    class RudderConfigBuilder
    {
        private string endPointUrl = Constants.BASE_URL;
        public RudderConfigBuilder WithEndPointUrl(string endPointUrl)
        {
            this.endPointUrl = endPointUrl;
            return this;
        }

        private List<RudderIntegrationFactory> factories = new List<RudderIntegrationFactory>();
        public RudderConfigBuilder WithFactory(RudderIntegrationFactory factory)
        {
            this.factories.Add(factory);
            return this;
        }

        public RudderConfigBuilder WithFactories(List<RudderIntegrationFactory> factories)
        {
            this.factories.AddRange(factories);
            return this;
        }

        private int flushQueueSize = Constants.FLUSH_QUEUE_SIZE;
        public RudderConfigBuilder WithFlushQueueSize(int flushQueueSize)
        {
            this.flushQueueSize = flushQueueSize;
            return this;
        }

        int dbCountThreshold = Constants.DB_COUNT_THRESHOLD;
        public RudderConfigBuilder WithDBCountThreshold(int dbCountThreshold)
        {
            this.dbCountThreshold = dbCountThreshold;
            return this;
        }

        int sleepTimeOut = Constants.SLEEP_TIME_OUT;
        public RudderConfigBuilder WithSleepTimeout(int sleepTimeOut)
        {
            this.sleepTimeOut = sleepTimeOut;
            return this;
        }

        int logLevel = RudderLogLevel.NONE;
        public RudderConfigBuilder WithLogLevel(int logLevel)
        {
            this.logLevel = logLevel;
            return this;
        }

        int configRefreshInterval = Constants.CONFIG_REFRESH_INTERVAL;
        public RudderConfigBuilder WithConfigRefreshInterval(int configRefreshInterval)
        {
            this.configRefreshInterval = configRefreshInterval;
            return this;
        }

        public RudderConfig Build()
        {
            return new RudderConfig(
                this.endPointUrl,
                this.flushQueueSize,
                this.dbCountThreshold,
                this.sleepTimeOut,
                this.logLevel,
                this.configRefreshInterval,
                this.factories
            );
        }
    }
}